woocommerce/woocommerce-ios
# [Mobile Payments] Add popular and last sold products on top of the products multi-selection screen

Closes: #9191 #9192 #9193 #9194


## Description

With this PR we add popular (most sold) and last sold products from cached orders on top of the Products Selection Screen on orders. Please note that since we delete all cached products every time we ask for the remotely, we can only show those top products that are included among those from the first page (25). If we load more, we show them on top and in the alphabetical list. Given the average products of Woo stores (41) I consider this feature can still provide good value to our users.

I also tried requesting the top products remotely so they can be shown even if they are not part of the first page, but given the latency of that request, it would defeat the purpose of Speedy IPP, to make the IPP collection process faster. You can see the long commit history to check all the task evolution.

Furthermore, I took the opportunity to rename some functions to be more in sync with the abstract purpose of those.

## Testing instructions


- Open app
- Go to Orders
- Tap on + (top right corner)
- Tap on + Add Products

- Please check that Popular (most sold) and last sold products are shown taken from your cached orders. Also, note that those products will be taken only from the first 25.

https://user-images.githubusercontent.com/1864060/234017866-94905504-0bcb-4d85-9358-807f341db5b1.mp4

Check that we don't display them when showing search or filters

https://user-images.githubusercontent.com/1864060/234018078-6cd4bfb9-f2a9-4354-ae00-6aad0bd12617.mp4

https://user-images.githubusercontent.com/1864060/234018116-e97ee52b-61d3-4913-9d97-a468be79dd7f.mp4

Please also check that the Product multi-selector on Coupons remains as before (no top products are shown)

## Known _issues_

- As remarked before, top products will be taken only from the cached ones. That is, on the first load 25, and then it will keep growing in parallel with the retrieval of the products.
- If you load all pages, all products are cached. When going to the Product selector again, we show the cached products before we load them remotely. With this, it might happen that we show a popular product from the n>1 pages briefly, only to delete it when the first 25 are loaded and cached. This happens because of the Product selector logic: we remove all cached products when loading new ones. As it is an edge case, I prefer to not make the code more complex/convoluted for this specific case. We can address it though if users get confused. Please let me know if you think otherwise.

---
